Bugfix/219 investigar falha aceitacao propostas

anthonysfarias opened the pull request:
## Melhorias no Tratamento de Erros 

### Mudanças Implementadas

#### 1. **Navegação Simplificada**
Antes o sistema fazia uma navegação complexa com redirecionamento temporário para a raiz e depois voltava para a página atual. Agora foi simplificado para ir diretamente para a página de gestão de licitações.

#### 2. **Tratamento Robusto de Erros**

Antes o sistema exibia a mensagem de erro errada, agora ele mostra a correta em caso de aceitar em lote. Aceitando a propósta em "Em análise", o sistema confirma e exibe a mensagem correta.

##### Parser Inteligente de Mensagens do Backend
- Suporte a múltiplos formatos de erro vindos do backend: arrays de erros, objetos com propriedade message, strings puras e JSON strings
- Sistema de fallback que usa mensagem padrão caso não consiga extrair mensagem específica
- Logging melhorado para facilitar debugging durante desenvolvimento

##### Detecção de Mensagem Específica
O sistema agora detecta mensagens específicas do backend, como a validação de lotes em análise, e exibe a tradução apropriada ou a mensagem original conforme necessário.

anthonysfarias committed:
fix(219): corrigir falha na aceitação de propostas

anthonysfarias committed:
fix: exibe a mensagem correta: Não é possível aceitar propostas enquanto o lote está em análise.

anthonysfarias committed:
fix: exibe a mensagem correta: Não é possível aceitar propostas enquanto o lote está em análise.

anthonysfarias committed:
fix: exibe a mensagem correta: Não é possível aceitar propostas enquanto o lote está em análise.

anthonysfarias committed:
Update environment.prod.ts

anthonysfarias committed:
Update environment.ts

juanmeleiro commented on src/app/project/proposal-screening/proposal-accepted/proposal-accepted.component.ts:
```diff
@@ -163,19 +193,42 @@ this.router.navigate([this.location.path()]);
       next: data => {
-        this.toastrService.success(this.translate.instant('TOASTRS.SUCCESS_ACCEPT_PROPOSAL'), '', { progressBar: true });
-        this.location.back();
+        this.toastrService.success('Proposta aceita com sucesso!', '', { progressBar: true });
```
Precisa usar o sistema de localização.

juanmeleiro reviewed (changes requested):
Só precisa trocar uma string que está hard-coded.

gabrielmussolini984 commented on src/app/project/proposal-screening/proposal-accepted/proposal-accepted.component.ts:
```diff
@@ -86,15 +86,45 @@ export class ProposalAcceptedComponent {
       error: error => {
-        console.error(error);
-        this.toastrService.error(this.translate.instant('TOASTRS.ERROR_ACCEPT_PROPOSAL'), '', { progressBar: true });
+        console.log('Erro completo:', error);
```
remover console.log

gabrielmussolini984 commented on src/app/project/proposal-screening/proposal-accepted/proposal-accepted.component.ts:
```diff
@@ -86,15 +86,45 @@ export class ProposalAcceptedComponent {
+      
+        if (backendMsg && typeof backendMsg === 'string' && backendMsg.length > 5) {
+          // Verifica se é a mensagem específica do lote em análise para traduzir
+          if (backendMsg.includes('Não é possível aceitar propostas enquanto o lote está em análise')) {
```
Usar dict

gabrielmussolini984 commented on src/app/project/proposal-screening/proposal-accepted/proposal-accepted.component.ts:
```diff
@@ -86,15 +86,45 @@ export class ProposalAcceptedComponent {
+          this.toastrService.error(this.translate.instant('TOASTRS.ERROR_ACCEPT_PROPOSAL'), '', { progressBar: true });
+        }
       }
     })
```
O backend tem que possuir um padrão de erro e qualquer throw que o backend retornar para o frontend vai ter a mesma estrutura, o máximo que o front vai fazer e ler essa estrutura e tomar decisões diferente(abrir um modal, ignorar, deixar a cor de mensagens diferente), porém a estrutura/tipagem sempre vai ser a msm.

anthonysfarias replied:
Simplifiquei o tratamento de erros no frontend, substituindo uma lógica bem extensa, com cerca de 30 linhas tentando adivinhar o formato do erro (arrays, objetos, strings, JSON etc.), por uma solução direta e enxuta de apenas 3 linhas. Como o backend já segue um padrão de estrutura para os erros, agora eu leio diretamente a mensagem retornada e uso uma mensagem localizada como fallback, caso necessário. Com isso, além de confiar mais na integração entre front e back, consegui reduzir bastante o código, deixar mais limpo e facilitar a manutenção, mantendo a consistência com o restante da aplicação

gabrielmussolini984 commented on src/app/project/proposal-screening/proposal-accepted/proposal-accepted.component.ts:
```diff
@@ -163,19 +193,42 @@ this.router.navigate([this.location.path()]);
       error: error => {
-        console.error(error);
-        this.toastrService.error(this.translate.instant('TOASTRS.ERROR_ACCEPT_PROPOSAL'), '', { progressBar: true });
+        console.log('Erro completo:', error);
+        let backendMsg = '';
+      
+        if (error?.error) {
+          // Caso o backend envie um array de erros
+          if (Array.isArray(error.error.errors) && error.error.errors.length > 0) {
+            backendMsg = error.error.errors[0];
+          }
+          // Caso o backend envie um objeto com campo message
+          else if (typeof error.error === 'object' && error.error.message) {
+            backendMsg = error.error.message;
+          }
+          // Caso o backend envie texto puro ou JSON string
+          else if (typeof error.error === 'string') {
+            try {
+              const parsed = JSON.parse(error.error);
+              backendMsg = parsed.message || error.error;
+            } catch {
+              backendMsg = error.error;
+            }
+          }
+        } else if (error?.message) {
+          backendMsg = error.message;
+        }
+      
+        if (backendMsg && typeof backendMsg === 'string' && backendMsg.length > 5) {
+          this.toastrService.error(backendMsg, '', { progressBar: true });
+        } else {
+          this.toastrService.error(this.translate.instant('TOASTRS.ERROR_ACCEPT_PROPOSAL'), '', { progressBar: true });
+        }
```
Mesma coisa

gabrielmussolini984 reviewed (changes requested):
Ajustar os comentários

anthonysfarias committed:
refactor(219): simplify error handling in proposal-accepted component

anthonysfarias committed:
refactor(219): simplify error handling in proposal-accepted component

anthonysfarias committed:
feat: add automatic backend error translation

- Add BackendErrorHandlerService with toast notifications
- Update BaseService to detect structured errors
- Add BACKEND_ERRORS translations (pt, en, es, fr)
- Complete end-to-end error mapping system

anthonysfarias commented:
O código ainda está em manutenção!

anthonysfarias committed:
feat: add i18n translations for structured error handling

anthonysfarias committed:
feat: add i18n translations for structured error handling

anthonysfarias committed:
Merge branch 'staging' into bugfix/219-investigar-falha-aceitacao-propostas

anthonysfarias commented:
@gabrielmussolini984 quando tiver um tempo, poderia revisar este PR? Agradeço desde já pelo feedback
Além disso, tivemos um problema de deploy, se puder dar uma olhada nisso também, eu agradeço!
